source/src/encryption/cs_ConnectionEncryption.cpp

```diff
@@ -144,6 +146,20 @@ void ConnectionEncryption::generateSessionData() {
 	RNG::fillBuffer(_sessionData.sessionNonce, sizeof(_sessionData.sessionNonce));
 	RNG::fillBuffer(_sessionData.validationKey, sizeof(_sessionData.validationKey));
 	memcpy(_nonce.sessionNonce, _sessionData.sessionNonce, sizeof(_sessionData.sessionNonce));
+	_log(LogLevelConnectionEncryption, false, "Set session data:");
+	_logArray(LogLevelConnectionEncryption, true, reinterpret_cast<uint8_t*>(&_sessionData), sizeof(_sessionData));
+}
+
+cs_ret_code_t ConnectionEncryption::setSessionData(session_data_t& sessionData) {
+	if (sessionData.protocol != _sessionData.protocol) {
+		return ERR_PROTOCOL_UNSUPPORTED;
+	}
+	memcpy(_sessionData.sessionNonce, sessionData.sessionNonce, sizeof(sessionData.sessionNonce));
+	memcpy(_sessionData.validationKey, sessionData.validationKey, sizeof(sessionData.validationKey));
+	memcpy(_nonce.sessionNonce, sessionData.sessionNonce, sizeof(sessionData.sessionNonce));
```

Comment:
If you pack nonce into a typed struct t, can also be a simple assignment:
```
struct packet_nonce_t {
    uint8_t data[PACKET_NONCE_LENGTH];
};
struct session_nonce_t { 
    uint8_t data[VALIDATION_KEY_LENGTH];
};
struct encryption_nonce_t{ 
...
}; 

// now _nonce.foo = someNonce; doesn't require memcpy anymore
```

Reply:
Isn't that a lot of effort to be able to change a memcpy into an assignment?